community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py

```diff
@@ -54,21 +55,86 @@
 
-NodeStatus = Enum(
-    "NodeStatus",
-    (
-        "orphan",
-        "power_down",
-        "preempted",
-        "restore",
-        "resume",
-        "terminated",
-        "unbacked",
-        "unchanged",
-        "unknown",
-    ),
-)
+class NodeAction(Protocol):
+    def apply(self, nodes:List[str]) -> None:
+        ...
+
+    def __hash__(self):
+        ...
+
+@dataclass(frozen=True)
+class NodeActionResume():
+    def apply(self, nodes:List[str]) -> None:
+        """resume nodes via scontrol"""
+        hostlist = util.to_hostlist_fast(nodes)
+        log.info(f"{len(nodes)} instances to resume ({hostlist})")
+        run(f"{lookup().scontrol} update nodename={hostlist} state=power_up")
+
+@dataclass(frozen=True)
+class NodeActionIdle():
+    def apply(self, nodes:List[str]) -> None:
+        """idle nodes"""
+        hostlist = util.to_hostlist_fast(nodes)
+        log.info(f"{len(nodes)} nodes to idle ({hostlist})")
+        run(f"{lookup().scontrol} update nodename={hostlist} state=resume")
+
+@dataclass(frozen=True)
+class NodeActionPowerDown():
+    def apply(self, nodes:List[str]) -> None:
+        """power_down node in slurm"""
+        hostlist = util.to_hostlist_fast(nodes)
+        log.info(f"{len(nodes)} instances to power down ({hostlist})")
+        run(f"{lookup().scontrol} update nodename={hostlist} state=power_down")
+
+@dataclass(frozen=True)
+class NodeActionDelete():
+    def apply(self, nodes:List[str]) -> None:
+        """delete instances for nodes"""
+        hostlist = util.to_hostlist_fast(nodes)
+        log.info(f"{len(nodes)} instances to delete ({hostlist})")
+        delete_instances(nodes)
+
+@dataclass(frozen=True)
+class NodeActionUnknown():
+    def apply(self, nodes:List[str]) -> None:
+        """Error status, nodes shouldn't get in this status"""
+        hostlist = util.to_hostlist_fast(nodes)
+        log.error(f"{len(nodes)} nodes have unexpected status: ({hostlist})")
+        first = next(iter(nodes))
+        state = lookup().slurm_node(first)
+        state = "{}+{}".format(state.base, "+".join(state.flags)) if state else "None"
+        inst = lookup().instance(first)
+        log.error(f"{first} state: {state}, instance status:{inst.status}")
+
+@dataclass(frozen=True)
+class NodeActionPrempt():
+    def apply(self, nodes:List[str]) -> None:
+        """Down nodes and then restart instances"""
+        NodeActionDown(reason="Preempted instance").apply(nodes)
+
+        hostlist = util.to_hostlist_fast(nodes)
+        log.info(f"{len(nodes)} instances restarted ({hostlist})")
+        start_instances(nodes)
 
+@dataclass(frozen=True)
+class NodeActionUnchanged():
+    def apply(self, nodes:List[str]) -> None:
+        pass
+
+@dataclass(frozen=True)
+class NodeActionDown():
+    reason: str
+
+    def apply(self, nodes: List[str]) -> None:
+        """down nodes"""
+        hostlist = util.to_hostlist_fast(nodes)
+        reason_quoted = shlex.quote(self.reason)
+        log.info(
```

Comment:
nit. too many lines

Reply:
Removed extra newlines and `reason_quoted`